### ISSUE TITLE
rendering issue

### ISSUE THREAD

**Issue by paul90** [fedwiki/wiki-plugin-activity#17]
Splitting the rendering issue raised in #16 by @opn to its own issue.

>There seems to be a plugin rendering / redraw issue - here is the recipe to replicate:
>1. Go to http://david.bovill.me/view/welcome-visitors
>2. Select [[My Cards]] Link
>3. Some times the Activity Plugin does no render the recent changes, if it does not click on the link above a second time.

>See this video - https://www.youtube.com/watch?v=rEcr12HQAhQ

and

>The activity plugin displays:
```
Page Activity
excluding neighborhood
```
>when displaying results from
```
ROSTER all
CONVERSATION
NARRATIVE 
```
>It would be better if instead of "excluding neighborhood" the plugin displayed "from roster"

**Comment by paul90** [comment]
An initial quick look:

1. The roster plugin is using the `REFERENCES david.bovill.me/davids-sites` to load the roster.
2. Looking at the network activity, on the occasions that the activity plugin doesn't do anything it would appear as if the roster get the references after the activity plugin is loaded.

The order that items get loaded does not appear to be consistent.

**Comment by WardCunningham** [comment]
We slow load site maps when they join the neighborhood. This involves events that keep activity aware of when it should refresh. We may just need to signal and existing event from the roster when activity completes.

**Comment by paul90** [comment]
While looking at fedwiki/wiki-client#158, which is related to activity around the `new-neighbor-done` event. I suspect that the problem here is that loading sites via REFERENCES in a ROSTER does not cause the `new-neighbor-done` event to be fired.

**Comment by WardCunningham** [comment]
The refresh cycle initiates sitemap loading once a new page has been displayed.
```
  whenGotten = (pageObject) ->
    buildPage( pageObject, $page )
    for site in pageObject.getNeighbors(location.host)
      neighborhood.registerNeighbor site
```
Here we are associating enlarging the neighborhood with display of pages. Plugins are not obligated to register anything about pages they fetch for their own purposes. Some other examples of plugins that fetch but don't register are Metabolism and Rollup, both early experiments with quietly fetching data from off-screen pages.

I don't think that Activity is under any obligation to add pages it fetches, or sites that those pages mention, to the neighborhood.

Activity should keep it self straight in the presence of neighborhood changes and its particular configurations. I've not explored this issue in advance of exacerbating it as noted here  https://github.com/fedwiki/wiki-client/pull/158#issuecomment-216620411 

I will explore. I am open to emitting an event other than new-neighbor-done when I modify the origin's already loaded sitemap. The existing sitemap is not new and is not technically a neighbor. However, I took the meaning of the event to be time-to-look-again-at-the-neighborhood-if-you-are-trying-to-represent-it.

**Comment by WardCunningham** [comment]
I will write some notes here regarding responding to incremental updates in the origin's sitemap.

The merge function constructs a list of "pages" to be displayed. These are rows of the report where information from like pages have been aggregated. Here is one of the rows.
```
[
  {
    "site": "garden.asia.wiki.org",
    "page": {
      "slug": "ward-cunningham",
      "title": "Ward Cunningham",
      "date": 1462370449094
    }
  },
  {
    "site": "localhost:3000",
    "page": {
      "slug": "ward-cunningham",
      "title": "Ward Cunningham",
      "date": 1440993473662
    }
  }
]
```
A good strategy might be for merge to delete duplicates as it merges fresher information from the recently revised origin sitemap.

**Comment by paul90** [comment]
I would be tempted to change that data structure from a list of maps to a map of maps keyed on the site,
``` json
{
  "garden.asia.wiki.org": {
    "site": "garden.asia.wiki.org",
    "page": {
      "slug": "ward-cunningham",
      "title": "Ward Cunningham",
      "date": 1462370449094
    }
  },
  "localhost:3000": {
    "site": "localhost:3000",
    "page": {
      "slug": "ward-cunningham",
      "title": "Ward Cunningham",
      "date": 1440993473662
    }
  }
}
```
it would then be a simple matter to replace the data for a particular site, rather than causing a duplicate, and the list of values from the map provides the same date as the list did previously.

**Comment by WardCunningham** [comment]
Yes, I agree. I tried making the `push` to the existing array conditional on an unsuccessful `find` hoping to demonstrate the simplicity of the solution.

I'm having a little trouble evaluating the fix because it was at the end of a series of experimental modifications that I may have had improperly installed or may have had async bugs of their own remaining. I set this work aside until the weekend when I wouldn't be hurried.
